Update for secure codec support

kenlsoft opened the pull request:
Hi,
This PR contains 5 commit to support real secure codec support.
clearkeydrmplugin now requires secure codec;
and created a standalone Aes decryptor for secure decrypt since there are many difference with original decryptor.
Rest commits are tiny to fix some warnings.

kenlsoft committed:
add .gitignore to ignore optee-clearkey-cdmi

Signed-off-by: Ken Liu <ken.liu@arm.com>

kenlsoft committed:
Purge compile warnings

Signed-off-by: Ken Liu <ken.liu@arm.com>

kenlsoft committed:
Use flag CFG_SECURE_DATA_PATH to trigger secure video path

Signed-off-by: Ken Liu <ken.liu@arm.com>

petegriffin commented on Android.mk:
```diff
@@ -17,6 +17,9 @@ LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
 
+# Extra flags for SDP usage
+USE_TEE_SDP ?= fasle
```
This should be CFG_SECURE_DATA_PATH to keep it consistent with other TA's like Playready (also it follows OP-TEE's nomenclature.

I think you can also make generating the secureDecryptor dependent on this flag

petegriffin commented on AesCtrDecryptorSecure.cpp:
```diff
@@ -0,0 +1,70 @@
+
+    if (_subSamples) {
+        uint32_t length = 0;
+        Iv opensslIv;
```
We can remove the references to openssl here

petegriffin commented on Android.mk:
```diff
@@ -17,8 +17,12 @@ LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
 
+# Extra flags for SDP usage
+USE_TEE_SDP ?= false
```
To be consistent with other TA's and also OP-TEE nomenclature we should enable this based on CFG_SECURE_DATA_PATH

petegriffin commented on CryptoPlugin.cpp:
```diff
@@ -41,9 +41,20 @@ ssize_t CryptoPlugin::decrypt(bool secure, const KeyId keyId, const Iv iv,
+    /* force secure mode */
+    ALOGI("%s: setting secure=1", __func__);
+    secure = 1;
+#elif USE_TEE_SDP==true
```
CFG_SECURE_DATA_PATH please

petegriffin commented on CryptoPlugin.h:
```diff
@@ -39,7 +39,12 @@ class CryptoPlugin : public android::CryptoPlugin {
 
     virtual bool requiresSecureDecoderComponent(const char* mime) const {
         UNUSED(mime);
+#if USE_TEE_SDP==true
```
CFG_SECURE_DATA_PATH please

petegriffin reviewed (changes requested):
Hi Ken,

Many thanks for your patches, and sorry for the delay in reviewing this pull request. I like the optimization of avoiding a REE to secure world transition for the clear data, and also of creating a new secure decryptor class.

As part of this work can you also update the unit test code in clearkeydrmplugin/tests/AesCtrDecryptorUnittest.cpp to instantiate a AesCtrDecryptorSecure class, and ensure those tests still pass.

Note for Android we aren't using the copy of this code which is also found in optee-clearkey-cdmi/tests directory.

KenLSoft commented:
Will update.

kenlsoft committed:
Init TEE context while session is created.

Signed-off-by: Ken Liu <ken.liu@arm.com>

KenLSoft commented:
Sorry I got some mis-operations. New patchsets below this comment would be the latest one.

For the test -- I can not find a good solution for the test since memory is protected by hardware ; and I need to remove the protection to check if result is correct; so I will leave test app later.

Other comments are all updated.

kenlsoft committed:
Add secure decryptor for secure frames

Secure frames combine with 5 bytes clear data and then
encrypted data. Send frame info into secure decryptor
and let it handle the memcpy and decryption instead of
trigger twice TEE call for each secure frame.

Signed-off-by: Ken Liu <ken.liu@arm.com>

petegriffin commented:
Hi Ken,

Thanks for the changes so far.

With regards to the test app and testing secure decryptor class. If you make the changes to instantiate the class and allocate a buffer from ION. I can test it on iMX8M-evk board as that platform has secure codecs support and I can disable the memory protection in ATF.

ajutras-linaro commented on Android.mk:
```diff
@@ -17,6 +17,9 @@ LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
 
+# Extra flags for SDP usage
+CFG_SECURE_DATA_PATH ?= fasle
```
To be consistent with other usages of CFG_SECURE_DATA_PATH within Linaro, this line should be:

+CFG_SECURE_DATA_PATH ?= n

To enable SDP, someone needs to export:

export CFG_SECURE_DATA_PATH=y

Otherwise, typo 'fasle' vs 'false'.

ajutras-linaro commented on CryptoPlugin.h:
```diff
@@ -39,7 +39,12 @@ class CryptoPlugin : public android::CryptoPlugin {
 
     virtual bool requiresSecureDecoderComponent(const char* mime) const {
         UNUSED(mime);
+#if CFG_SECURE_DATA_PATH==true
```
To also be consistent with other usages of CFG_SECURE_DATA_PATH within Linaro, CFG_SECURE_DATA_PATH would only be defined when it is enabled. At this line, #ifdef CFG_SECURE_DATA_PATH would be used.

KenLSoft commented:
Thanks for reviewing.
@petegriffin I tried to write the test but seeems I need to call platform APIs for secure buffer allocating instead of an output buffer in stack; which means needs some code change for it; so I planned to create another PR for the test. What's your option? Thanks.

petegriffin commented:
Hi Ken,

Sorry for the long delay between in replies. I'm actually on parental leave atm (OoO July & August). It would be preferable to update the unitary test code at the same time as adding the new secure decryptor class so we can be sure we've not introduced any regressions.

KenLSoft commented:
OK.
And since I am fully occupied these days so that my response won't be that quick, sorry for that.

petegriffin commented:
Hi Ken,

I just thought..I think you could pretty much re-use the ion implementation Alex did here https://github.com/linaro-mmwg/open-content-decryption-module/commit/1a8870e93ba4e83a72424478ec05b2882cae4753 and add it to the unit test code to allocate the secure buffer.

KenLSoft commented:
Ah. that code is useful; is it ok to copy one or two files from that project?

petegriffin commented:
Yes it is Linaro written code, so we can re-use it here

kenlsoft committed:
Add AesCtrDecryptorSecureUnittest

Class AesCtrDecryptorSecure test purpose. Need to shutdown
hardware firewall to pass the test; or there would be a
bus error reported while testing.

Signed-off-by: Ken Liu <ken.liu@arm.com>

ajutras-linaro commented:
The changes look good to me.
